packages/foam-core/src/note-graph.ts

```diff
@@ -1,121 +1,162 @@
-    frontmatter: object,
-    title: string | null,
-    links: NoteLink[],
-    definitions: NoteLinkDefinition[],
-    start: Point,
-    end: Point,
-    path: string,
-    source: string,
-    eol: string
-  ) {
-    this.id = id;
-    this.frontmatter = frontmatter;
-    this.title = title;
-    this.source = source;
-    this.path = path;
-    this.links = links;
-    this.definitions = definitions;
-    this.start = start;
-    this.end = end;
-    this.eol = eol;
-  }
-}
-
 export type NoteGraphEventHandler = (e: { note: Note }) => void;
 
+export type NoteQuery = {
+  id?: ID;
+  slug?: string;
+  uri?: URI;
+  title?: string;
+};
+
 export class NoteGraph {
   private graph: Graph;
   private events: EventEmitter;
+  private createIdFromURI: (uri: URI) => ID;
 
   constructor() {
     this.graph = new Graph();
     this.events = new EventEmitter();
+    this.createIdFromURI = hashURI;
   }
 
-  public setNote(note: Note) {
-    const noteExists = this.graph.hasNode(note.id);
+  public setNote(note: NoteInfo) {
+    const id = this.createIdFromURI(note.source.uri);
+    const slug = uriToSlug(note.source.uri);
+    const noteExists = this.graph.hasNode(id);
     if (noteExists) {
-      (this.graph.outEdges(note.id) || []).forEach(edge => {
+      (this.graph.outEdges(id) || []).forEach(edge => {
         this.graph.removeEdge(edge);
       });
     }
-
-    this.graph.setNode(note.id, note);
+    const graphNote: Note = {
+      ...note,
+      id: id,
+      slug: slug,
+    };
+    this.graph.setNode(id, graphNote);
     note.links.forEach(link => {
-      const slugger = new GithubSlugger();
-      this.graph.setEdge(note.id, slugger.slug(link.to), link.text);
+      const relativePath =
+        note.definitions.find(def => def.label === link.slug)?.url ?? link.slug;
+      const targetPath = getUriViaRelative(note.source.uri, relativePath);
+      const targetId = this.createIdFromURI(targetPath);
+      const connection: Connection = {
+        from: graphNote.id,
+        to: targetId,
+        link: link,
+      };
+      this.graph.setEdge(graphNote.id, targetId, connection);
     });
-
-    this.events.emit(noteExists ? 'update' : 'add', { note });
+    this.events.emit(noteExists ? 'update' : 'add', { note: graphNote });
+    return graphNote;
   }
 
   public getNotes(): Note[] {
-    return this.graph.nodes().map(id => this.graph.node(id));
+    return this.graph
+      .nodes()
+      .map(id => this.graph.node(id))
+      .filter(Boolean);
   }
 
-  public getNote(noteId: ID): Note | void {
-    if (this.graph.hasNode(noteId)) {
-      return this.graph.node(noteId);
+  public getNoteId(query: NoteQuery): ID | null {
+    if (query.id) {
+      return query.id;
+    }
+    const searchFn = query.uri
+      ? (note: Note | null) => note?.source.uri === query.uri
+      : query.slug
+      ? (note: Note | null) => note?.slug === query.slug
+      : query.title
+      ? (note: Note | null) => note?.title === query.title
```

Comment:
Are titles unique in the workspace? This seems like it could have clashes?

(Sorry if you saw and dismissed this comment earlier. I left it before but I think it got nuked as outdated by a push)

Reply:
`id` and `uri` are unique. `slug` and `title` could have 0+ results. in case there is > 1 result, the first one is returned.
I needed something for `slug` and I thought it could be done is such a way as to allow to query the graph.
some comments:
- if we want to call out the difference between picking in the graph or scanning it, we could have 2 fns. I don't think that's necessary as it feels like an implementation detail (and we could improve the performance of various searches by adding maps)
- we could have a `listNotes(query: NoteQuery)` that returns an array for those situations where we actually want all the results (e.g. autocomplete)

of course we could also simply have `getByUri` and `getBySlug` (or `listBySlug` as they are not unique)

Reply:
The bit I don't like here is that it's too easy to accidentally link to the wrong note due to slug/title collision. Enshrining their non-uniqueness to the API would help avoid these accidents from future implementers.

I'm not prissy about the naming.